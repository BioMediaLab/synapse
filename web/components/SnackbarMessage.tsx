import React, { Component } from "react";
import {
  IconButton,
  Snackbar,
  SnackbarContent,
  withStyles,
} from "@material-ui/core";
import { green, amber } from "@material-ui/core/colors";
import CheckCircleIcon from "@material-ui/icons/CheckCircle";
import ErrorIcon from "@material-ui/icons/Error";
import InfoIcon from "@material-ui/icons/Info";
import WarningIcon from "@material-ui/icons/Warning";
import CloseIcon from "@material-ui/icons/Close";

const styles = theme => ({
  error: {
    backgroundColor: theme.palette.error.dark,
  },
  success: {
    backgroundColor: green[600],
  },
  info: {
    backgroundColor: theme.palette.primary.dark,
  },
  warning: {
    backgroundColor: amber[700],
  },
  icon: {
    fontSize: 20,
    opacity: 0.9,
    marginRight: theme.spacing.unit,
  },
  message: {
    display: "flex",
    alignItems: "center",
  },
});

interface ISnackbarMessageProps {
  message: any;
  success?: boolean;
  error?: boolean;
  info?: boolean;
  warning?: boolean;
  classes: {
    error: string;
    success: string;
    info: string;
    warning: string;
    icon: string;
    message: string;
  };
}

interface ISnackbarMessageState {
  open: boolean;
}

class SnackbarMessage extends Component<
  ISnackbarMessageProps,
  ISnackbarMessageState
> {
  state = {
    open: true,
  };

  componentDidUpdate(prevProps) {
    if (prevProps.message != this.props.message) {
      this.setState({ open: true });
    }
  }

  handleClose = event => {
    this.setState({ open: false });
  };

  render() {
    const props = this.props;
    const classes = props.classes;

    var snackbarContent = <div />;
    if (!props.message) return null;
    else {
      if (props.error) {
        snackbarContent = (
          <SnackbarContent
            message={
              <span id="client-snackbar" className={classes.message}>
                <ErrorIcon className={classes.icon} />
                {props.message.message
                  ? props.message.message.replace("GraphQL error: ", "")
                  : props.message.replace("GraphQL error: ", "")}
              </span>
            }
            className={classes.error}
            action={[
              <IconButton
                key="close"
                aria-label="Close"
                color="inherit"
                onClick={this.handleClose}
              >
                <CloseIcon className={classes.icon} />
              </IconButton>,
            ]}
          />
        );
      } else {
        snackbarContent = (
          <SnackbarContent
            message={
              <span id="client-snackbar" className={classes.message}>
                {props.success ? (
                  <CheckCircleIcon className={classes.icon} />
                ) : props.info ? (
                  <InfoIcon className={classes.icon} />
                ) : props.warning ? (
                  <WarningIcon className={classes.icon} />
                ) : null}
                {props.message}
              </span>
            }
            className={
              props.success
                ? classes.success
                : props.info
                ? classes.info
                : props.warning
                ? classes.warning
                : null
            }
            action={[
              <IconButton
                key="close"
                aria-label="Close"
                color="inherit"
                onClick={this.handleClose}
              >
                <CloseIcon className={classes.icon} />
              </IconButton>,
            ]}
          />
        );
      }
      return (
        <Snackbar
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "left",
          }}
          open={this.state.open}
          autoHideDuration={6000}
          onClose={this.handleClose}
        >
          {snackbarContent}
        </Snackbar>
      );
    }
  }
}

export default withStyles(styles)(SnackbarMessage);
