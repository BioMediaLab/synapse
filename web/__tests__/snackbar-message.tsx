import React from "react";
import { render } from "react-testing-library";
import SnackbarMessage from "../components/SnackbarMessage";

const graphqlErrorObject = { message: "GraphQL error: graphql error message" };

test("renders", () => {
  const { container } = render(
    <>
      <SnackbarMessage message="message" />
      <SnackbarMessage error message="error message" />
      <SnackbarMessage error message={graphqlErrorObject} />
      <SnackbarMessage success message="success message" />
      <SnackbarMessage info message="info message" />
      <SnackbarMessage warning message="warning message" />
    </>,
  );
  expect(container).toMatchSnapshot();
});
