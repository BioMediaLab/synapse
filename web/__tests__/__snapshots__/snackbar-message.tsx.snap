// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div
    class="MuiSnackbar-root-7 MuiSnackbar-anchorOriginBottomLeft-13"
  >
    <div
      class="MuiTypography-root-44 MuiTypography-body1-53 MuiPaper-root-17 MuiPaper-elevation6-25 MuiSnackbarContent-root-14"
      role="alertdialog"
      style="-webkit-transform: translate(0, 0); transform: translate(0, 0); -webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message-15"
      >
        <span
          class="SnackbarMessage-message-6"
          id="client-snackbar"
        >
          message
        </span>
      </div>
      <div
        class="MuiSnackbarContent-action-16"
      >
        <button
          aria-label="Close"
          class="MuiButtonBase-root-86 MuiIconButton-root-80 MuiIconButton-colorInherit-81"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-85"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-98"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiSnackbar-root-7 MuiSnackbar-anchorOriginBottomLeft-13"
  >
    <div
      class="MuiTypography-root-44 MuiTypography-body1-53 MuiPaper-root-17 MuiPaper-elevation6-25 MuiSnackbarContent-root-14 SnackbarMessage-error-1"
      role="alertdialog"
      style="-webkit-transform: translate(0, 0); transform: translate(0, 0); -webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message-15"
      >
        <span
          class="SnackbarMessage-message-6"
          id="client-snackbar"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
            />
          </svg>
          error message
        </span>
      </div>
      <div
        class="MuiSnackbarContent-action-16"
      >
        <button
          aria-label="Close"
          class="MuiButtonBase-root-86 MuiIconButton-root-80 MuiIconButton-colorInherit-81"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-85"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-98"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiSnackbar-root-7 MuiSnackbar-anchorOriginBottomLeft-13"
  >
    <div
      class="MuiTypography-root-44 MuiTypography-body1-53 MuiPaper-root-17 MuiPaper-elevation6-25 MuiSnackbarContent-root-14 SnackbarMessage-error-1"
      role="alertdialog"
      style="-webkit-transform: translate(0, 0); transform: translate(0, 0); -webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message-15"
      >
        <span
          class="SnackbarMessage-message-6"
          id="client-snackbar"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
            />
          </svg>
          graphql error message
        </span>
      </div>
      <div
        class="MuiSnackbarContent-action-16"
      >
        <button
          aria-label="Close"
          class="MuiButtonBase-root-86 MuiIconButton-root-80 MuiIconButton-colorInherit-81"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-85"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-98"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiSnackbar-root-7 MuiSnackbar-anchorOriginBottomLeft-13"
  >
    <div
      class="MuiTypography-root-44 MuiTypography-body1-53 MuiPaper-root-17 MuiPaper-elevation6-25 MuiSnackbarContent-root-14 SnackbarMessage-success-2"
      role="alertdialog"
      style="-webkit-transform: translate(0, 0); transform: translate(0, 0); -webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message-15"
      >
        <span
          class="SnackbarMessage-message-6"
          id="client-snackbar"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          success message
        </span>
      </div>
      <div
        class="MuiSnackbarContent-action-16"
      >
        <button
          aria-label="Close"
          class="MuiButtonBase-root-86 MuiIconButton-root-80 MuiIconButton-colorInherit-81"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-85"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-98"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiSnackbar-root-7 MuiSnackbar-anchorOriginBottomLeft-13"
  >
    <div
      class="MuiTypography-root-44 MuiTypography-body1-53 MuiPaper-root-17 MuiPaper-elevation6-25 MuiSnackbarContent-root-14 SnackbarMessage-info-3"
      role="alertdialog"
      style="-webkit-transform: translate(0, 0); transform: translate(0, 0); -webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message-15"
      >
        <span
          class="SnackbarMessage-message-6"
          id="client-snackbar"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
            />
          </svg>
          info message
        </span>
      </div>
      <div
        class="MuiSnackbarContent-action-16"
      >
        <button
          aria-label="Close"
          class="MuiButtonBase-root-86 MuiIconButton-root-80 MuiIconButton-colorInherit-81"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-85"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-98"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiSnackbar-root-7 MuiSnackbar-anchorOriginBottomLeft-13"
  >
    <div
      class="MuiTypography-root-44 MuiTypography-body1-53 MuiPaper-root-17 MuiPaper-elevation6-25 MuiSnackbarContent-root-14 SnackbarMessage-warning-4"
      role="alertdialog"
      style="-webkit-transform: translate(0, 0); transform: translate(0, 0); -webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message-15"
      >
        <span
          class="SnackbarMessage-message-6"
          id="client-snackbar"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
            />
          </svg>
          warning message
        </span>
      </div>
      <div
        class="MuiSnackbarContent-action-16"
      >
        <button
          aria-label="Close"
          class="MuiButtonBase-root-86 MuiIconButton-root-80 MuiIconButton-colorInherit-81"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-85"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-89 SnackbarMessage-icon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-98"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
